pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -56,42 +57,65 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	ctx = ctrl.LoggerInto(ctx, log)
 	log.V(2).Info("Reconciling StatefulSet")
 
-	err = r.fetchAndFinalizePods(ctx, req.Namespace, req.Name)
+	err = r.fetchAndFinalizePods(ctx, sts)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) fetchAndFinalizePods(ctx context.Context, namespace, statefulSetName string) error {
+func (r *Reconciler) fetchAndFinalizePods(ctx context.Context, sts *appsv1.StatefulSet) error {
 	podList := &corev1.PodList{}
-	if err := r.client.List(ctx, podList, client.InNamespace(namespace), client.MatchingLabels{
-		pod.GroupNameLabel: GetWorkloadName(statefulSetName),
+	if err := r.client.List(ctx, podList, client.InNamespace(sts.Namespace), client.MatchingLabels{
+		podcontroller.GroupNameLabel: GetWorkloadName(sts.Name),
 	}); err != nil {
 		return err
 	}
-	return r.finalizePods(ctx, podList.Items)
+	return r.finalizePods(ctx, sts, podList.Items)
 }
 
-func (r *Reconciler) finalizePods(ctx context.Context, pods []corev1.Pod) error {
+func (r *Reconciler) finalizePods(ctx context.Context, sts *appsv1.StatefulSet, pods []corev1.Pod) error {
 	log := ctrl.LoggerFrom(ctx)
 	return parallelize.Until(ctx, len(pods), func(i int) error {
-		p := &pods[i]
-		if p.Status.Phase != corev1.PodSucceeded && p.Status.Phase != corev1.PodFailed {
-			return nil
-		}
-		err := clientutil.Patch(ctx, r.client, p, true, func() (bool, error) {
-			removed := controllerutil.RemoveFinalizer(p, pod.PodFinalizer)
-			if removed {
-				log.V(3).Info("Finalizing pod in group", "pod", klog.KObj(p), "group", p.Labels[pod.GroupNameLabel])
+		pod := &pods[i]
+		return client.IgnoreNotFound(clientutil.Patch(ctx, r.client, pod, true, func() (bool, error) {
+			if finalizePod(sts, pod) {
+				log.V(3).Info(
+					"Finalizing pod in group",
+					"pod", klog.KObj(pod),
+					"group", pod.Labels[podcontroller.GroupNameLabel],
+				)
+				return true, nil
 			}
-			return removed, nil
-		})
-		return client.IgnoreNotFound(err)
+			return false, nil
+		}))
 	})
 }
 
+func finalizePod(sts *appsv1.StatefulSet, pod *corev1.Pod) bool {
+	var updated bool
+
+	if shouldUngate(sts, pod) && utilpod.Ungate(pod, podcontroller.SchedulingGateName) {
+		updated = true
+	}
```

Comment:
FYI: When a Pod has a scheduling gate, it should be ungated because StatefulSet does not allow updating Pods with a scheduling gate.

Reply:
sgtm